### PR TITLE
Extend the schizo framework

### PR DESCRIPTION
**Pull request by rhc54**
Allow definition of CLI options by environment. Refactor orterun to mesh with the orted_submit code, thus improving code reuse. Eliminate the orte-submit tool as orterun can now meet that need.

### CODE REVIEW

**Comment by hppritcha** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by rhc54** [comment]
I was seeing some weirdness from github in general - took me several tries to get the updated branch to appear in my repo. So these failures may just be another symptom of something going on with github today.

**Comment by rhc54** [comment]
bot:retest

File: orte/mca/plm/base/plm_base_launch_support.c [hunk 1]
```diff
@@ -1534,11 +1534,15 @@ int orte_plm_base_setup_virtual_machine(orte_job_t *jdata)
...
     }
+    if (NULL == daemons->map) {
+        daemons->map = OBJ_NEW(orte_job_map_t);
+    }
     map = daemons->map;
 
     /* if this job is being launched against a fixed DVM, then there is
      * nothing for us to do - the DVM will stand as is */
     if (orte_get_attribute(&jdata->attributes, ORTE_JOB_FIXED_DVM, NULL, OPAL_BOOL)) {
+        opal_output(0, "DVM");
```
Comment (jjhursey) [inline orte/mca/plm/base/plm_base_launch_support.c]: I bet this line was leftover debugging. Probably needs to be removed.

File: orte/mca/plm/base/plm_base_launch_support.c [hunk 2]
```diff
@@ -1559,16 +1562,16 @@ int orte_plm_base_setup_virtual_machine(orte_job_t *jdata)
...
-            ++(daemons->map->num_nodes);
+            opal_pointer_array_add(map->nodes, (void*)node);
+            ++(map->num_nodes);
             /* maintain accounting */
             OBJ_RETAIN(node);
             /* mark that this is from a singleton */
             singleton = true;
         }
-        map = daemons->map;
+        opal_output(0, "SPAWNED");
```
Comment (jjhursey) [inline orte/mca/plm/base/plm_base_launch_support.c]: Extra debugging here too.

File: orte/mca/plm/base/plm_base_launch_support.c [hunk 3]
```diff
@@ -1614,17 +1617,8 @@ int orte_plm_base_setup_virtual_machine(orte_job_t *jdata)
      * all nodes with application procs on them
      */
     if (orte_get_attribute(&daemons->attributes, ORTE_JOB_NO_VM, NULL, OPAL_BOOL)) {
+        opal_output(0, "NOVM");
```
Comment (jjhursey) [inline orte/mca/plm/base/plm_base_launch_support.c]: ++debugging

File: orte/mca/rmaps/base/rmaps_base_map_job.c [hunk 4]
```diff
@@ -532,6 +462,7 @@ void orte_rmaps_base_map_job(int fd, short args, void *cbdata)
             }
             free(output);
         }
+        exit(1);
```
Comment (jjhursey) [inline orte/mca/rmaps/base/rmaps_base_map_job.c]: Just want to confirm that we want to call `exit` here instead of a shutdown mechanism.

Comment (rhc54) [inline orte/mca/rmaps/base/rmaps_base_map_job.c]: Yeah, these are some leftovers. I created a "Work-In-Progress, Do-Not-Merge!" label and marked this with it. I'm trying to let some other folks poke at it to see if it does what they need, then they will augment and we need to cleanup prior to merge.

Still, I do appreciate the review! So please feel free to poke at it. I know there are several places that still require cleanup.

**Comment by hppritcha** [comment]
What does the label stand for?

**Comment by rhc54** [comment]
It means this is "Work-In-Progress, Do-Not-Merge". I've asked some folks to check the branch to ensure it provides the basic operations they were asking for, and then it needs some cleanup before it will be ready. This PR just lets us track that the changes aren't breaking anything (at least, per Jenkins) while we work on it, and gives others a "heads up".

**Comment by rhc54** [comment]
Let me provide more clarity as to what we are attempting to do so others can perhaps contribute or at least review with context. The objective of this effort is to allow tools to "mimic" a variety of environments. As an example, we might want mpiexec to take the options currently employed by Platform MPI so that users transitioning from Platform to OMPI will not have to modify their scripts.

These changes allow the selected component to define the command line options as they like, and then they will parse the actual command line against that definition.

Prior changes provide the ability to "stack" these definitions, so the framework stubs will cycle across the components in priority order to allow them to:

- detect if they were specifically identified, or if they detect that we are in an environment where they should be used;
- add their command line option definitions to the overall definition

So you could create a tool, for example, that has common options regardless of environment, and then have schizo add the environment-specific options.

HTH

**Comment by rhc54** [comment]
Not sure what this PR has to do with the mindist mapper, so let's try this again

**Comment by rhc54** [comment]
bot:retest

**Comment by rhc54** [comment]
looks like there is a problem on the mellanox jenkins server - getting random errors, this time a memory allocation problem in the tcp btl.

**Comment by rhc54** [comment]
Verified with MTT:

```shell
+-------------+-----------------+-------------+----------+------+------+----------+------+--------------------------------------------------------------------------+
| Phase       | Section         | MPI Version | Duration | Pass | Fail | Time out | Skip | Detailed report                                                          |
+-------------+-----------------+-------------+----------+------+------+----------+------+--------------------------------------------------------------------------+
| MPI Install | my installation | 3.0.0a1     | 00:01    | 1    |      |          |      | MPI_Install-my_installation-my_installation-3.0.0a1-my_installation.html |
| Test Build  | trivial         | 3.0.0a1     | 00:00    | 1    |      |          |      | Test_Build-trivial-my_installation-3.0.0a1-my_installation.html          |
| Test Build  | ibm             | 3.0.0a1     | 00:40    | 1    |      |          |      | Test_Build-ibm-my_installation-3.0.0a1-my_installation.html              |
| Test Build  | intel           | 3.0.0a1     | 01:14    | 1    |      |          |      | Test_Build-intel-my_installation-3.0.0a1-my_installation.html            |
| Test Build  | onesided        | 3.0.0a1     | 00:04    | 1    |      |          |      | Test_Build-onesided-my_installation-3.0.0a1-my_installation.html         |
| Test Build  | java            | 3.0.0a1     | 00:02    | 1    |      |          |      | Test_Build-java-my_installation-3.0.0a1-my_installation.html             |
| Test Build  | orte            | 3.0.0a1     | 00:01    | 1    |      |          |      | Test_Build-orte-my_installation-3.0.0a1-my_installation.html             |
| Test Run    | trivial         | 3.0.0a1     | 00:06    | 6    |      |          |      | Test_Run-trivial-my_installation-3.0.0a1-my_installation.html            |
| Test Run    | ibm             | 3.0.0a1     | 09:32    | 380  |      |          |      | Test_Run-ibm-my_installation-3.0.0a1-my_installation.html                |
| Test Run    | spawn           | 3.0.0a1     | 00:09    | 6    | 1    |          |      | Test_Run-spawn-my_installation-3.0.0a1-my_installation.html              |
| Test Run    | loopspawn       | 3.0.0a1     | 10:02    | 1    |      |          |      | Test_Run-loopspawn-my_installation-3.0.0a1-my_installation.html          |
| Test Run    | intel           | 3.0.0a1     | 24:07    | 474  |      |          | 4    | Test_Run-intel-my_installation-3.0.0a1-my_installation.html              |
| Test Run    | intel_skip      | 3.0.0a1     | 12:08    | 431  |      |          | 47   | Test_Run-intel_skip-my_installation-3.0.0a1-my_installation.html         |
| Test Run    | onesided        | 3.0.0a1     | 01:30    | 30   |      | 2        |      | Test_Run-onesided-my_installation-3.0.0a1-my_installation.html           |
| Test Run    | java            | 3.0.0a1     | 00:00    | 1    |      |          |      | Test_Run-java-my_installation-3.0.0a1-my_installation.html               |
| Test Run    | orte            | 3.0.0a1     | 00:45    | 19   |      |          |      | Test_Run-orte-my_installation-3.0.0a1-my_installation.html               |
+-------------+-----------------+-------------+----------+------+------+----------+------+--------------------------------------------------------------------------+


    Total Tests:    1358
    Total Failures: 3
    Total Passed:   1355
    Total Duration: 3621 secs. (01:00:21)

```

**Pull request merged**